### PR TITLE
PoC: Optimise SSR under heavy load

### PR DESCRIPTION
**Pull request by dunqan**
POC for #8755. 

There are a few things needed to make this production ready, but this is a good start.

### CODE REVIEW

File: projects/storefrontapp/optimized-ssr-engine.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  timeout?: number;
+
+  /**
+   * Enable in-memory cache for pre-rendered urls
+   */
+  cache?: boolean;
+}
+
+export function optimizedSsrEngine(
+  originalEngine,
```
Comment (tobi-or-not-tobi) [inline projects/storefrontapp/optimized-ssr-engine.ts]: Rename to `expressEngine` ?

File: projects/storefrontapp/optimized-ssr-engine.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  ssrOptions?: OptimizedSsrOptions
+) {
+  const renderedUrls: {
```
Comment (tobi-or-not-tobi) [inline projects/storefrontapp/optimized-ssr-engine.ts:21]:
```suggestion
  // The rendered pages are kept in memory to be served on next request. If the `cache` is set to `false`, the 
  // response is evicted as soon as the first successful response is successfully returned. 
  const renderedUrls: {
```

File: projects/storefrontapp/optimized-ssr-engine.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import * as fs from 'fs';
+
+export interface OptimizedSsrOptions {
+  /**
+   * Time in milliseconds to wait for SSR rendering to happen.
+   */
+  timeout?: number;
+
+  /**
+   * Enable in-memory cache for pre-rendered urls
```
Comment (tobi-or-not-tobi) [inline projects/storefrontapp/optimized-ssr-engine.ts]:
```suggestion
   * Enable in-memory cache for pre-rendered urls.
```

File: projects/storefrontapp/optimized-ssr-engine.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  return function wrappedEngine(
+    filePath: string,
+    options: any,
+    callback: (err?: Error | null, html?: string) => void
+  ) {
+    const res = options.res || options.req.res;
+
+    /**
+     * When SSR page can not be returned in time, we're returnig index.html of
+     * the CDR application.
```
Comment (tobi-or-not-tobi) [inline projects/storefrontapp/optimized-ssr-engine.ts]:
```suggestion
     * the CSR application.
```

File: projects/storefrontapp/optimized-ssr-engine.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+     * The CSR application will have "Cache-Control: no-store" header to avoid caching of the fallback.
```
Comment (tobi-or-not-tobi) [inline projects/storefrontapp/optimized-ssr-engine.ts]:
```suggestion
     * The CSR application is returned with the "Cache-Control: no-store" response-header. This notifies external cache systems to not use the CSR application for the subsequent request. 
```

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]

File: projects/storefrontapp/optimized-ssr-engine.ts (hunk 1 above)
Comment (dunqan) [reply projects/storefrontapp/optimized-ssr-engine.ts]: 👍

**Review by tobi-or-not-tobi — APPROVED** [review]

**Pull request closed without merging**